### DISCUSSION TITLE
delay() vs signal/event

### DISCUSSION THREAD

**Discussion by earnubs** [mswjs/msw#1307] [Ideas]
We found that some of our tests were flaky depending on environment with the standard 5ms `delay()` timeout for node environments, so we came up with:

```js
test('loading state shows loading button', async () => {
    const hold = new EventEmitter();

    server.use(
      graphql.mutation(
        'UpdateThing',
        async (req, res, ctx) => {
          // we want to hold back the response until we are ready, rather than
          // try to guess a delay value
          return new Promise((resolve) => {
            hold.once('release', () => {
              resolve(
                res(
                  ctx.errors([
                    {
                      // make this a standard response so it's reusable,
                      // we don't actally care what it is
                      message: 'loading test only', 
                    },
                  ])
                )
              )
            })
          });
        }
      )
    );
    const { user } = renderThing();

    const saveButton = await screen.findByRole('button', { name: 'Save' });
    
    await user.click(saveButton);

    // Button should be loading now
    await waitFor(async () => {
      expect(screen.getByRole('button', { name: 'Loading' })).toBeInTheDocument();
    });

    // let the msw response promise resolve, so test exits cleanly
    hold.emit('release'); 
  });

What do you think?

**Comment by kettanaito** [comment]
Hey, @earnubs. Thanks for raising this discussion. 

I think this falls under #778, as you should be testing loading states using `ctx.delay('infinite')`. This will make sure that the request remains pending for the entirety of your test, and you can assert the loading state of your UI with confidence. As the linked issue states, there's another challenge to that, as infinite request will keep pending, and there's no logic right now to halt it once the test ends. 

If you have any ideas on how we could reliably do that, please share them in that issue, I'd highly appreciate your involvement! 

Analyzing your approach, I wouldn't hold the response resolution as a part of your test. These kind of criteria must live in MSW, and that's where `ctx.delay('infinite')` should be applied. Think of it this way: your test represents the scenario that the user will experience. But the user won't manually hold the resolution of requests, right? So neither should the test. MSW, on the other hand, is here to describe _server-side_ behavior, and that's the natural place to control how long the request remains pending.

**Reply by lortimer** [reply]
@kettanaito I hear your argument that a test should model a realistic scenario and @earnubs' idea violates that. In an ideal situation, you're absolutely right. I came here because I was trying something similar to earnubs, and I couldn't find a way to both return the response from the `ResponseResolver` _and_ delay the response until a signal from my test.

I think the problem described by #778 is far worse than the consequences of compromising on our ideals. To me, `ctx.delay('infinite')` is unusable in its current form. As a team lead, I will not allow warnings and error messages to pollute my test output, because that inevitably leads to [alert fatigue](https://www.atlassian.com/incident-management/on-call/alert-fatigue) and developers ignoring all kinds of error output.

Earnubs' idea is a good solution to #778, and is consistent with other JavaScript testing tools. RxJS observables are a staple of Angular, and they allow a developer to construct a scenario in which the test controls resolution of a network request. Jasmine and Jest both pass a function called `done` to test methods so a developer can run a timer in the test and call `done()` at the end so the test runner knows the test didn't time out.

Here are a couple of suggestions for how tests using msw might work if you think this kind of solution is acceptable to solve #778. I'm sure there is a more user-friendly way to do it than I can think of right now. I'm very happy to help implement something like this.

1. Pass an event emitter and event name to `ctx.delayUntil`
```tsx
let emitter = new EventEmitter();

server.use(
  rest.post("endpoint", (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.body({message: "success!"},
      ctx.delayUntil(emitter, "event name") 
    )
  })
);

afterEach(() => {
  emitter.emit("event name"); // emit event in afterEach in case a test did not
});

test('should ... while request is pending', () => {
  render(<Component/>);
  expect(screen.getByText("Loading...").toBeVisible();
});

test('should ... when request resolves', () => {
  render(<Component/>);
  emitter.emit("event name");
  expect(screen.getByText("Loading...").not.toBeVisible();
});
```

2. ctx.delayUntil provides a `done` function via a callback
```tsx

let resolveRequest: () => void;
server.use(
  rest.post("endpoint", (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.body({message: "success!"},
      ctx.delayUntil((done) => {resolveRequest = done}) 
    )
  })
);

afterEach(() => {
  resolveRequest(); // clean up in afterEach in case a test did not call done
});

test('should ... while request is pending', () => {
  render(<Component/>);
  expect(screen.getByText("Loading...").toBeVisible();
});

test('should ... when request resolves', () => {
  render(<Component/>);
  resolveRequest();
  expect(screen.getByText("Loading...").not.toBeVisible();
});
```

**Reply by kettanaito** [reply]
The great part is that you can implement both variants of `delayUntil` even now, right in your code base:

```js
async function delayUntil(emitter, event) {
  return new Promise((resolve) => {
    emitter.once(event, resolve)
  })
}

rest.post('/endpoint', async (req, res, ctx) => {
  await delayUntil(emitter, 'done')
  return res(ctx.text('Hello world'))
})

afterEach(() => {
  emitter.emit('done')
})
```

But it's one thing to cover this in a single codebase and a totally different thing to design an API that hundreds of codebases would be encouraged to use. It comes down to a good internal/external control balance, as I don't wish for developers to spin up emitters just to implement infinite request pending logic. This is a good example of implementation detail leakage, and I'm convinced that features like this one should be covered by the library. 

I think one of the smartest suggestions I've heard is to rely on `server.close()` to terminate any pending side-effects, such as infinite delays that are `setTimeout()` under the hood. I don't see why this wouldn't work if you call `server.close()` in your `afterAll` hook, which you should be doing anyway.
